### PR TITLE
Implements partial inflation minting to treasury

### PR DESCRIPTION
**Pull request by gpestana**
This PR implements an optional minting of the inflation directly into a pot.

The staking pallet now has a percent storage item, `InflationLevyFraction`, which defines the percentage of the era inflation that should be minted directly into an account (the treasury's account). The remaining of the fraction is the validators rewards for that era. The `InflationLevyFraction` can be set via extrinsic by the staking admin origin through the `set_treasury_fraction` callable. The `InflationLevyFraction` default is 0. 

#### Treasury `Get<AccountId>` implementation
This PR also implements the `Get<AccountId>` in the treasury pallet which exposes an account id derived at the origin. This helps the staking pallet to request the account ID of the treasury pallet, instead of deriving it locally based on its pallet ID.

Closes https://github.com/paritytech/polkadot-sdk/issues/403

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3635571

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3639281

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3640608

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-additional-tests
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3641266

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-additional-tests
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3641681

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-additional-tests
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3641825

**Comment by gpestana** [comment]
bot bench substrate-pallet --pallet=pallet_staking

**Comment by gpestana** [comment]
bot bench polkadot-pallet --runtime polkadot

**Comment by gpestana** [comment]
bot clean

File: substrate/frame/staking/src/pallet/mod.rs [hunk 1]
```diff
@@ -1776,6 +1785,15 @@ pub mod pallet {
 			MinCommission::<T>::put(new);
 			Ok(())
 		}
+
+		/// Sets the minimum amount of commission that each validators must maintain.
```
Comment (franciscoaguirre) [inline substrate/frame/staking/src/pallet/mod.rs:1789]:
```suggestion
		/// Sets the fraction of the era's inflation that should go to the treasury account
```

File: substrate/frame/staking/src/pallet/mod.rs [hunk 2]
```diff
@@ -580,6 +583,10 @@ pub mod pallet {
 	#[pallet::storage]
 	pub(crate) type ChillThreshold<T: Config> = StorageValue<_, Percent, OptionQuery>;
 
+	/// The percentage of the reward inflation that is minted directly into the treasury.
+	#[pallet::storage]
+	pub type TreasuryInflationFraction<T: Config> = StorageValue<_, Percent, OptionQuery>;
```
Comment (franciscoaguirre) [inline substrate/frame/staking/src/pallet/mod.rs:588]: With `ValueQuery` you get the default value which should be `Zero::zero()` all the time.

**Review by kianenigma — COMMENTED** [review]
Somewhat skeptical of this implementation being in the right path.

Staking already have a way to communicate "what should I do with the rest of my rewards?", which is `RewardRemainder`. 

It seems a bit sloppy to add another similar mechanism that says "before I distribute my rewards, what should I do with it?".

I kinda wish there was a way to reuse `RewardRemainder` to bundle it into a single new hook that expresses what should be done with rewards. 

Lastly, you are making a few important assumptions: 

1. That _a treasury_ pallet exists.
2. _How_ this treasury pallet's pallet id should be converted into an account id. 

Ideally, all of this should be more abstract.

**Comment by gpestana** [comment]
> Staking already have a way to communicate "what should I do with the rest of my rewards?", which is `RewardRemainder`.
My idea is that the treasury payment and validator rewards are both first citizen reward "receivers", thus there could be still a  remainder. I will look into this direction again though.
 
> Ideally, all of this should be more abstract.
Good point, I think that there are a few things that could be changed: 1) renaming "TreasuryPot" to something else, 2) have a way to request the account ID of a type that implements a new `Pot` trait that exposes a way to get an account_id. This last point is already implemented in the last commit. 

Using the trait `Pot` also helps making the implementation safer, since it offloads the burden of deriving the account ID to the origin so that we don't need to keep the account ID derivation in 2 different places, which may be error-prone in case of updates.

File: substrate/frame/staking/src/pallet/mod.rs [hunk 3]
```diff
@@ -197,6 +197,9 @@ pub mod pallet {
 		/// See [Era payout](./index.html#era-payout).
 		type EraPayout: EraPayout<BalanceOf<Self>>;
 
+		/// The treasury pot account ID.
+		type TreasuryPalletId: Get<frame_support::PalletId>;
```
[the remark this answers is not in the document]
Comment (gpestana) [reply substrate/frame/staking/src/pallet/mod.rs]: I changed the way that the account ID of the treasury/pot is derived. So this comment may be addressed already.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3671945

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3671916

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3672150

**Comment by gpestana** [comment]
@kianenigma to touch base on your comment to abstract this further and use the `RewardRemainder` as a way to pay the treasury fraction:

My current perspective is that the `RewardRemainder` should be the remainder of the total payout (validator rewards *and* treasury %), based on the yearly inflation and the total payouts.

Following that, I think that having an explicit `InflationLevyFraction` and `InflationLevyDestination = Get<AccountId>`, and explicitly use those types to calculate and mint directly into the treasury is a reasonable way to implement this. wdyt?

Otoh, we could refactor the way we handle the reward remainder and `RewardRemainder` to both handle the inflation remainder and the treasury payment, but I think that's not conceptually the best way (i.e. the treasury payout is not a "remainder").

**Comment by gpestana** [comment]
Being replaced by https://github.com/paritytech/polkadot-sdk/pull/1660

**Pull request closed without merging**